File: day_2/day_2.py
from pathlib import Path


def parse_input():
    input_file = Path(__file__).parent / "input.txt"
    return [line.strip() for line in input_file.open()]


def part1(commands: list[str]):
    submarine = {"position": 0, "depth": 0}
    for command in commands:
        match command.split():
            case ["forward", value]:
                submarine["position"] += int(value)
            case ["up", value]:
                submarine["depth"] -= int(value)
            case ["down", value]:
                submarine["depth"] += int(value)
            case _:
                print(command)
    return submarine["depth"]*submarine["position"]

def part2(commands: list[str]):
    submarine = {"position": 0, "depth": 0, "aim": 0}
    for command in commands:
        match command.split():
            case ["forward", value]:
                submarine["position"] += int(value)
                submarine["depth"] += submarine["aim"] * int(value)
            case ["up", value]:
                submarine["aim"] -= int(value)
            case ["down", value]:
                submarine["aim"] += int(value)
            case _:
                print(command)
    return submarine["depth"]*submarine["position"]

def main():
    data = parse_input()
    print(f"{part1(data)=}")
    print(f"{part2(data)=}")


if __name__ == "__main__":
    main()

File: day_2/day_2.nim
import std/[strutils, strformat, sugar, tables]
from std/os import parentDir

proc parse_input(): seq[string] =
  const file = currentSourcePath.parentDir() & "\\input.txt"
  result = collect(newSeq):
    for line in file.lines: line.strip

proc part1(commands: seq[string]): int =
  var submarine = to_table({"position": 0, "depth": 0})
  for command in commands:
    let cmd = command.split
    case cmd[0]:
      of "forward":
        submarine["position"] += cmd[1].parseInt
      of "up":
        submarine["depth"] -= cmd[1].parseInt
      of "down":
        submarine["depth"] += cmd[1].parseInt
  result = submarine["position"] * submarine["depth"]

proc part2(commands: seq[string]): int =
  var submarine = to_table({"position": 0, "depth": 0, "aim": 0})
  for command in commands:
    let cmd = command.split
    case cmd[0]:
      of "forward":
        submarine["position"] += cmd[1].parseInt
        submarine["depth"] += submarine["aim"] * cmd[1].parseInt
      of "up":
        submarine["aim"] -= cmd[1].parseInt
      of "down":
        submarine["aim"] += cmd[1].parseInt
  result = submarine["depth"]*submarine["position"]

proc main(): void =
  let data = parse_input()
  echo(fmt"{part1(data)=}")
  echo(fmt"{part2(data)=}")

when is_main_module:
  main()
